tensorizer/serialization.py

```diff
@@ -4446,17 +4128,492 @@ def write_state_dict(self, state_dict: Union[Dict, List, Tuple]):
+                )
+                with torch.cuda.stream(self._stream):
+                    staging_tensor_view.copy_(write_spec.tensor)
+                write_spec.user_owns_tensor_data = False
+                write_spec.tensor = staging_tensor_view.clone().detach()
+
+        max_tensor_size = max(
+            [w.tensor.element_size() * w.tensor.nelement() for w in cuda_specs]
+        )
+        cuda_transfer = CudaTransfer(max_tensor_size)
+
+        for w in cuda_specs:
+            assert w.tensor_data_task is None
+            w.tensor_data_task = cuda_transfer.submit(w)
+            self._jobs.append(w.tensor_data_task)
+
+        return cuda_transfer.executor
+
+    def _prepare_for_write_contiguous(self, write_specs: Sequence[_WriteSpec]):
+        def make_contiguous(write_spec, dependency):
+            if dependency is not None:
+                dependency.result(_TIMEOUT)
+            write_spec.tensor = write_spec.tensor.contiguous()
+            write_spec.data_length = write_spec.tensor.nbytes
+            write_spec.user_owns_tensor_data = False
+
+        for w in write_specs:
+            # if there is a tensor_data_task it is a cuda tensor
+            if w.tensor_data_task is not None or w.tensor.is_contiguous():
+                continue
+            w.tensor_data_task = self._computation_pool.submit(
+                make_contiguous, w, w.tensor_data_task
+            )
+
+    def _prepare_for_write_dtype(self, write_specs: Sequence[_WriteSpec]):
+        torch_dtype_to_numpy_dtype_cache: Dict[str, str] = {}
+
+        for w in write_specs:
+            tensor_dtype_str = str(w.tensor.dtype)
+            if _NumpyTensor._is_asymmetric(w.tensor.dtype):
+                # is opaque
+                w.dtype = (
+                    f"<V{w.tensor.element_size():d}"
+                    + OPAQUE_DTYPE_SEP
+                    + tensor_dtype_str
+                )
+                w.set_min_file_version_number(OPAQUE_TENSORIZER_VERSION)
+            else:
+                w.dtype = torch_dtype_to_numpy_dtype_cache.get(
+                    tensor_dtype_str, None
+                )
+                if w.dtype is None:
+                    torch_dtype_to_numpy_dtype_cache[tensor_dtype_str] = (
+                        w.dtype
+                    ) = (
+                        torch.tensor((), dtype=w.tensor.dtype, device="cpu")
+                        .numpy()
+                        .dtype.str
+                    )
+
+    @staticmethod
+    def _do_clone(write_spec, dependency: Optional[_Future]):
+        if dependency is not None:
+            dependency.result(_TIMEOUT)
+        write_spec.tensor = write_spec.tensor.clone().detach()
+        write_spec.numpy_tensor = _NumpyTensor.from_tensor(write_spec.tensor)
+
+    def _prepare_for_write_encryption(
+        self, write_specs: Sequence[_WriteSpec]
+    ) -> None:
+        assert self._encrypted and self._encryption is not None
```

Comment:
I'd try to avoid `assert` statements for anything regarding encryption correctness (or anything else important) since `assert` statements can be skipped when optimization is enabled, just in case.
```suggestion
        if not self._encrypted or self._encryption is None:
            raise RuntimeError(
                "Tried to encrypt tensors without encryption parameters"
                " having been provided"
            )
```